client/src/features/visualizer-threejs/ConfigControls.tsx

```diff
@@ -0,0 +1,245 @@
+        return null;
+    }
+
+    return (
+        <div className={"controls-container"}>
+            <div className="controls__list">
+                {inputs.map((i) => {
+                    return (
+                        <div key={i.key} className="controls__item">
+                            <label>{i.label}</label>
+                            <input
+                                type="number"
+                                value={visualizerConfigValues[i.key]}
+                                onChange={(e) => handleChange(i.key, e.target.value)}
+                            />
+                            {!!errors[i.key] && <div className={"controls__error"}>{errors[i.key]}</div>}
+                        </div>
+                    );
+                })}
+            </div>
+
+            <div className="controls__actions">
+                <button type={"button"} onClick={handleApply}>
+                    Apply
+                </button>
+                {showResetButton && (
+                    <button
+                        onClick={() => {
+                            localStorage.removeItem(VISUALIZER_CONFIG_LOCAL_STORAGE_KEY);
+                            setShowResetButton(false);
+                            location.reload();
+                        }}
+                    >
+                        Reset
+                    </button>
+                )}
+            </div>
+
+            <div style={{ marginTop: "16px" }}>
+                <label style={{ display: "block" }}>Zoom</label>
+                <input
+                    value={localZoom === undefined ? "" : localZoom}
+                    onChange={(e) => {
+                        const input = e.target.value;
+
+                        if (!input) {
+                            setLocalZoom(undefined);
+                            return;
+                        }
+
+                        const numberRegExp = /^-?\d+(\.|\.\d*|\d*)?$/;
+                        if (numberRegExp.test(input)) {
+                            if (input.endsWith(".")) {
+                                setLocalZoom(input as any);
+                            } else {
+                                const value = parseFloat(input);
+                                if (value > 2) {
+                                    setLocalZoom(2);
+                                    return;
+                                }
+                                setLocalZoom(value);
+                            }
+                        }
+                    }}
+                />
+                <div className="controls__actions">
+                    <button
+                        type={"button"}
+                        onClick={() => {
+                            if (localZoom === undefined || `${localZoom}`.endsWith(".")) {
+                                return;
+                            }
+                            setForcedZoom(localZoom);
```

Comment:
Same here

Reply:
I undertand. Then I suggest changing the type for the use state to string, since you need to actually store a string to show the value, and then only set when it's converted correctly to number. This is because type assertion is not generally a good practice, unless needed, and here, you are actually skipping the type check and storing a string in a number typed state `setLocalZoom(input as any);`

Reply:
Just this one more change and it's good to go I think 👍